106 commonly used ses variables

shijiaSMH opened the pull request:
### Closes issue 

#106
### Major changes: 
- Add a function to merge all commonly used socioeconomic (SES) variables to a cohort input. 
- Add a vignette to provide user basic understanding about variable concepts & how to use

### Additional information, if applicable:
For reviewer, can you pls review function (`nbhd_ses_variables.R`) & vignette (`nbhd_ses_variables.Rmd`) for clarity? 
- For vignette, it's a good sign that you feel confident using all variables after reading. Can you ensure all reference links work properly?
- For function, no concerns so far.

Saved vignette output here: `R:\GEMINI\Research Projects\Rgemini\nbhd_ses_variables`

I didn't do unit test since function is about merging.

Use case: Expect analyst to use `nbhd_ses_variables.R` function to grab all the SES variables needed, and use `nbhd_ses_variables.Rmd` as a reference to obtain basic understanding of those variables.

loffleraSMH commented on R/nbhd_ses_variables.R:
```diff
@@ -0,0 +1,121 @@
+                               
+    ) %>% as.data.table()
+    
+    # cal % visible minority
+    var_tbl[, c21_vismin_pct := round(c21_vismin / (c21_vismin + c21_vismin_not) * 100, 2 )]
+    
+    # cal % immigration status
+    var_tbl[, c21_immsta_pct := round(c21_immsta_imm / c21_immsta * 100, 2 )]
+    
+    # cal % post-secondary education (>15)
+    var_tbl[, c21_ed_15over_postsec_pct := round(c21_ed_15over_postsec / c21_ed_15over * 100, 2 )]
+    
+    # cal % post-secondary education (25-64)
+    var_tbl[, c21_ed_25to64_postsec_pct := round(c21_ed_25to64_postsec / c21_ed_25to64 * 100, 2 )]
+    
+    var_tbl <- var_tbl[, !c('c21_vismin', 'c21_vismin_not', 'c21_immsta_imm', 'c21_immsta', 'c21_ed_15over_postsec', 'c21_ed_15over', 'c21_ed_25to64_postsec', 'c21_ed_25to64'
+    )] 
```
This section is a little bit redundant because the calculations are the same as for the 2016 census data. The only thing that's different is the prefix (`c21` vs. `c16`) in the variable names, so you could potentially consider calculating this outside of the if statement.

loffleraSMH reviewed (changes requested):
Thanks so much for adding this @shijiaSMH.

For this first round of review, I focused on the function `nbhd_ses_variables()`. I checked the queries and calculations in the function and made sure the returned outputs are as expected. There are a few things below that need to be addressed (e.g., see comments about DB version, handling of missing values, and documentation).

Once we’ve finalized this, let’s reevaluate the need for a vignette. Typically, vignettes are used to illustrate specific use cases for functions, especially if there are a lot of input arguments (like in the cell suppression/plotting functions) or if the calculation within the function is complex (e.g., readmission). I don’t think either of these criteria apply here, so I’m not sure the vignette is needed. All the documentation of inputs/outputs and relevant resources should be in the roxygen instead (see comments below).

Please also add the new function to `_pkgdown.yml` and also mention it in `NEWS.md` (see guidelines in [CONTRIBUTING.md](https://github.com/GEMINI-Medicine/Rgemini/blob/master/CONTRIBUTING.md)).

I will do another round of review after you’ve had a chance to address my first round of comments.

shijiaSMH commented:
Thanks @loffleraSMH! I've moved the vignette documentation to the function itself. The new function has repetition but I think the logic is clear. Would happy to shorten the codes if we believe there's a solution for quick implementation. 

I've added the new function to `_pkgdown.yml`, mention it in `NEWS.md`, and `NAMESPACE`

Pls close the thread if you think my comment/edit is sufficient, and let me know if you have other suggestions!

shijiaSMH reviewed (commented):
Thanks @loffleraSMH! I've moved the vignette documentation to the function itself. The new function has repetition but I think the logic is clear. Would happy to shorten the codes if we believe there's a solution for quick implementation. 

I've added the new function to `_pkgdown.yml`, mention it in `NEWS.md`, and `NAMESPACE`

Function name was changed to improve read-ability, based on Surain's suggestion. 

Pls close the thread if you think my comment/edit is sufficient, and let me know if you have other suggestions!

loffleraSMH commented on R/nbhd_ses_variables.R:
```diff
@@ -0,0 +1,121 @@
+#' Version of dissemination_area chosen. Only 2016 or 2021 are valid inputs. 
+#' 
+#' @param cohort\cr
+#' User specified cohort
```
Nice, can you also add the expected class for `cohort`? Ideally, the function accepts both `data.table` and `data.frame`:
```
@param cohort (`data.table` | `data.frame`)\cr
```

In this case, you'll need to adjust the code to make sure it actually works if cohort is provided as a `data.frame` (at the moment, it results in an error). You can fix that by adding the following to the beginning of your code (right before you write the temp table):
```
check_input(cohort, c("data.table", "data.frame"), colnames = "genc_id")
cohort <- coerce_to_datatable(cohort)
```

shijiaSMH replied:
implemented

loffleraSMH commented on R/nbhd_ses_variables.R:
```diff
@@ -0,0 +1,121 @@
+                                          
+                                          left join locality_variables l
+                                            on l.genc_id = tmp.genc_id
+                                          left join lookup_statcan_v2016 s
```
Nice, I tested the function on `report_db_v3` and `drm_clean_db_v2_1_3` and it now works fine on both versions. The only small suggestion I have is to reword the warning message as follows:
```
Your DB version only contains 2016 census data, please change your input to `census_year = 2016`
```

shijiaSMH replied:
implemented

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#' Obtain commonly used neighbourhood-level socioeconomic (SES) variables
+#'
+#' @description
+#' The `neighborhood_ses()` function merges SES variables based on dissemination area ID. Function descriptions contains the necessary understanding of variable concepts and corresponding application recommendations.
```
implemented

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#' @section Data source
+#' All SES variables are sourced from [Statistics Canada Census.](https://www12.statcan.gc.ca/census-recensement/2021/ref/98-304/2021001/chap1-eng.cfm)
+#'
+#' The Census of Population provides a detailed and comprehensive statistical portrait of Canada. It’s collected by Statistics Canada conducts once every 5 years, aiming to provide population and dwelling counts for communities of all sizes across Canada.
```
I shortened it even further, pls that that out in edits.

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#'
+#' On-Marg is a neighborhood-level index showing marginalization differences between areas. It’s created using a portion of Statistics Canada Census variables.
+#' On-Marg has 4 dimensions. There was a name change in 2021 to avoid deficit-based language and better reflect the census measures associated with each dimension
+#' Summary score can be used if all dimensions are in the same direction. If >1 dimensions are in opposite direction, it’s recommended to report those dimensions separately.
```
made them at the end in `@section Ontario Marginalization Index (On-Marg)`

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#' On-Marg has 4 dimensions. There was a name change in 2021 to avoid deficit-based language and better reflect the census measures associated with each dimension
+#' Summary score can be used if all dimensions are in the same direction. If >1 dimensions are in opposite direction, it’s recommended to report those dimensions separately.
+#'
+#' @section Income additional details
```
removed

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#' - Ontario Marginalization Index (numeric): (instability_da16, deprivation_da16, dependency_da16, ethniccon_da16) or (households_dwellings_DA21, material_resources_DA21, age_labourforce_DA21, racialized_NC_pop_DA21)
+#' - Ontario Marginalization Index (quintile): (instability_q_da16, deprivation_q_da16, dependency_q_da16, ethniccon_q_da16) or (households_dwellings_q_DA21, material_resources_q_DA21, age_labourforce_q_DA21, racialized_NC_pop_q_DA21)
+#'
+#' @references
+#' [ON-Marg guide 2016](https://www.publichealthontario.ca/-/media/documents/U/2018/userguide-on-marg.pdf)
+#' [ON-Marg guide 2021](https://www.publichealthontario.ca/-/media/Event-Presentations/2023/09/ontario-marginalization-index-updates-products.pdf?rev=07baae2569164c17abaa18464075aa20&sc_lang=en)
```
added

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#' The Census of Population provides a detailed and comprehensive statistical portrait of Canada. It’s collected by Statistics Canada conducts once every 5 years, aiming to provide population and dwelling counts for communities of all sizes across Canada.
+#'
+#' On-Marg is a neighborhood-level index showing marginalization differences between areas. It’s created using a portion of Statistics Canada Census variables.
+#' On-Marg has 4 dimensions. There was a name change in 2021 to avoid deficit-based language and better reflect the census measures associated with each dimension
```
added description of each dimension, note 1: 2016 vs 2021 description seems similar, so those description may seem repetitive, but it's hard to aggregate them since variable names were changed with a legit reason. 

Note 2: There's room for argument for whether to put 2016 or 2021 first. As a counter argument: many people are using an older version of DB since their research started long ago, making them having to stick w 2016. Personally feel neutral about which version to go first.

shijiaSMH commented on R/nbhd_ses_variables.R:
```diff
@@ -0,0 +1,121 @@
+  }
+  
+  ## message to show invalid meanings of null values
+  message('Null values or 9 or 999 represent unknown or missing estimates')
```
Document an unexpected case where income quintile=999 (income continuous=999999) have known DA in 2021 data. See codes below to replicate. Based on this result, they represent encounter with DA available but missing income. Will aggregate those to 'DA available but census/ON Marg missing'.

For 2016 data, income quintile=9 (income continuous=99999999) have all DA ending in 9999, we can safely believe those are unknown DA. Would aggregate those as 'Missing DA'.

```
source('/mnt/nfs/home/shijia/GitHub/Rgemini/R/utils.R')
source('/mnt/nfs/home/shijia/GitHub/Rgemini/R/neighborhood_ses.R')

drv <- dbDriver("PostgreSQL") 
dbcon <- DBI::dbConnect(drv,
  dbname = "report_db_v3_9_1", # change DB to load later
  host = "prime.smh.gemini-hpc.ca",
  port = 5432,
  user = "shijia",
  password = getPass::getPass("password")
)

### check 2021 data
var_tbl2 <- DBI::dbGetQuery(dbcon, 'select tmp.genc_id, l.da21uid, s.c21_vismin, s.c21_vismin_not, s.qnatippe, s.qnbtippe, s.qaatippe, s.qabtippe, s.atippe, s.btippe, s.c21_immsta, s.c21_immsta_imm, s.c21_ed_15over_postsec, s.c21_ed_15over, s.c21_ed_25to64_postsec, s.c21_ed_25to64, "households_dwellings_DA21", "material_resources_DA21", "age_labourforce_DA21", "racialized_NC_pop_DA21", "households_dwellings_q_DA21", "material_resources_q_DA21", "age_labourforce_q_DA21", "racialized_NC_pop_q_DA21"

                                          from temp_data tmp

                                          left join locality_variables l
                                            on l.genc_id = tmp.genc_id
                                          left join lookup_statcan_v2021 s
                                            on l.da21uid = s.da21uid; ') %>% as.data.table()
tst <- var_tbl2[!is.na(da21uid)]
tst[qnatippe==999]$da21uid %>% unique
 [1] 46230062 10090080 48150089 35510100 10090107 10090124 35390757 35370839 35600308 10090117 10090079 48040059
```

shijiaSMH replied:
It's unexpected in the sense that income=9 actually has valid DA in 2021 data. This affects my codes, but we are still using 2 types of missingness. See edits for details later

loffleraSMH replied:
Please re-review the information I shared above. What you describe refers to the 2nd type of missingness I mentioned - this is very much expected.

Your current code is still not correct. Why do you assume DA is invalid if `qnatippe == 9`, and why do you only make this assumption for 2016 data? As mentioned above, PCCF explicitly mentions the values referring to invalid DAs. Please use that in your code. The unique DAs you get with your code above are NOT invalid DAs. Again, they are valid DAs but with missing income data (see 2nd type of missingness).

loffleraSMH replied:
Hi @shijiaSMH: This is still not working as expected. Why are you applying this `if` statement when cleaning up missing values? `(census_year != 2016 & tnl != 1)`. This means that none of this code (including the warning messages) are executed when the user runs this function with `census_year = 2016` on the new DB version with 2 statcan tables. Please carefully check the function logic and test it with different user inputs to make this works as intended. 
In this context: I would also highly recommend removing some redundancy in the code, so it's easier to catch bugs like this.

shijiaSMH replied:
Hi @loffleraSMH, thanks for catching this. Now that I think of it, we have 2 options: 1) keep this message, not clean up 9s in income; 2) remove this message, clean up 9s in income.

Which seems better? 

It's quite strange about this one: (census_year != 2016 & tnl != 1). This means that none of this code (including the warning messages) are executed when the user runs this function with census_year = 2016 on the new DB version with 2 statcan tables.

I tested on both clean DB and report DB, things all run normally. Thanks for flagging, I'll test again.

loffleraSMH replied:
Hi @shijiaSMH: Please see above, we want to do both - 1) first, set 9s in income to `NA` since they represent missing values **AND** 2) show a warning message about the final % of `NA` (after setting 9s to `NA`).

If `census_year = 2016` and 2 tables are found (e.g., in `report_db_v3`) the `if` statement returns `FALSE`. Also, why would this `if` statement be relevant here at all? The clean-up/warning messages should be run regardless of the specific input arguments/DB version. Let's make sure the code logic is clear here.

shijiaSMH replied:
Hi @loffleraSMH, thanks again for those suggestions. I believe all those points are addressed in the update.

Note. I the function ran normally on `report_db_v3_9_1`, `drm_cleandb_v2_1_3`, `drm_cleandb_v3_0_0`. I conducted the tests 2 times. 

Please keep me posted if anything doesn't work from your end.

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+  rec <- var_tbl %>%
+    select(all_of(coln_keep))
+
+  # change to final names
+  setnames(rec, "vm", paste0("c", yr, "_vismin_pct"))
+  setnames(rec, "im", paste0("c", yr, "_immsta_pct"))
+  setnames(rec, "ed15", paste0("c", yr, "_ed_15over_postsec_pct"))
+  setnames(rec, "ed25", paste0("c", yr, "_ed_25to64_postsec_pct"))
```
removed the year component in those output variables, have simplified the codes

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+  setnames(rec, "ed15", paste0("c", yr, "_ed_15over_postsec_pct"))
+  setnames(rec, "ed25", paste0("c", yr, "_ed_25to64_postsec_pct"))
+
+  if (census_year == 2016) {
+    # cal % exclusion for 2016 census data
+    p1 <- nrow(rec[qnatippe == 9]) / nrow(rec)
+    p2 <- nrow(rec[is.na(get(paste0("da", yr, "uid")))]) / nrow(rec)
+    p3 <- nrow(rec[!is.na(get(paste0("da", yr, "uid"))) & (
+      is.na(qnatippe) |
+        is.na(instability_da16) |
+        is.na(get(paste0("c", yr, "_vismin_pct"))) |
+        is.na(get(paste0("c", yr, "_immsta_pct"))) |
+        is.na(get(paste0("c", yr, "_ed_15over_postsec_pct"))) |
+        is.na(get(paste0("c", yr, "_ed_25to64_postsec_pct"))))]) / nrow(rec)
+
+    message(paste0(round(p1, 4), "% encounters are excluded due to missing income information during census data collection"))
+    message(paste0(round(p2, 4), "% encounters are excluded due to missing dissemination area"))
+    message(paste0(round(p3, 4), "% encounters are excluded due to no census data linkage with dissemination area"))
+    # delete missing data
+    rec <- na.omit(rec)
+    rec <- rec[qnatippe != 9]
+  }
```
edited

shijiaSMH commented on R/neighborhood_ses.R:
```diff
@@ -0,0 +1,191 @@
+#' This function returns a data.table where each row corresponds to a genc_id from the user-provided cohort input, together with the following columns:
+#' - DA (dissemination area) ID from the corresponding census year: da16uid or da21uid. DA is ['a small area composed of one or more neighbouring dissemination blocks and is the smallest standard geographic area for which all census data are disseminated.'](https://www150.statcan.gc.ca/n1/en/catalogue/92-169-X)
+#' - [Visible minority](https://www12.statcan.gc.ca/census-recensement/2016/ref/dict/pop127-eng.cfm): c16_vismin_pct or c21_vismin_pct
+#' - [Neighbourhood-level income](https://www.cihi.ca/sites/default/files/document/toolkit-area-level-measurement-pccf-en.pdf): qnatippe, qnbtippe, qaatippe, qabtippe, atippe, btippe. While qnatippe and qnbtippe are quintiles calculated based on national income distribution, while qaatippe and qabtippe are quintiles constructed separately for each census metropolitan area (CMA), census agglomerat ...
```
I updated the text section about Statistics Canada Census of Population. There are some intricacies about income and immigration, pls see text for details. 

I mentioned long-format questionnaire, I think the 25% households info is enough to hint those variables are less reliable. I don't want to explicitly mention this since Stats Can argues those are representative sample geographically.

shijiaSMH reviewed (commented):
Hi @loffleraSMH, I believe I've addressed all your comments, let me know if further edits are needed!

loffleraSMH commented:
Closing this PR here since I refactored the function and will create a new PR for it shortly.
